client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,314 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::{collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
+use std::time::Instant;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber, Level, span::{Attributes, Record, Id}};
+use tracing_subscriber::CurrentSpan;
+
+use sc_client_api::BlockBackend;
+use sc_rpc_server::MAX_PAYLOAD;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+// Default to only events with the key prefixes for :extrinsic_index & system::Account.
+const DEFAULT_STORAGE_KEYS: &'static str =
+	"3a65787472696e7369635f696e646578,26aa394eea5630e07c48ae0c9558cef7b99d880ec681799c0cf30e8886371da9";
+
+struct BlockSubscriber {
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Mutex<HashMap<Id, SpanDatum>>,
+	events: Mutex<Vec<TraceEvent>>,
+}
+
+impl BlockSubscriber {
+	fn new(
+		targets: &str,
+		spans: Mutex<HashMap<Id, SpanDatum>>,
+		events: Mutex<Vec<TraceEvent>>,
+	) -> Self {
+		let next_id = AtomicU64::new(1);
+		let mut targets: Vec<_> = targets
+			.split(',')
+			.map(|s| crate::parse_target(s))
+			.collect();
+		// Ensure that WASM traces are always enabled
+		// Filtering happens when decoding the actual target / level
+		targets.push((WASM_TRACE_IDENTIFIER.to_owned(), Level::TRACE));
+		BlockSubscriber {
+			targets,
+			next_id,
+			current_span: CurrentSpan::default(),
+			spans,
+			events,
+		}
+	}
+}
+
+impl Subscriber for BlockSubscriber {
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				return true;
+			}
+		}
+		false
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let id = self.next_id.fetch_add(1, Ordering::Relaxed);
```

Comment:
@mattrutherford Just read the `Atomics` docs for the first time so bear with me ... with `Ordering::Relaxed` it does not seem like Ids are guaranteed to be in the order the span is entered. Can we change this too `Ordering::AcqRel`? 
From what I understand it should be minimal added cost on x86 platforms (bc x86 is strongly ordered) and then we can remove the sort and double collect for spans since we should be guaranteed that Id is incremented in the order `new_span` is invoked.

Reply:
It's fine to "unordered" it just used as a simple counter and the ordered should not matter, Basti explained this very nicely [here](https://github.com/paritytech/substrate/pull/7780#discussion_r620558709)

However, as we talked about earlier it's easy to fuck it up if we decide that change that during refactoring or something. So better to wrap the `AtomicU64` is a newtype that only allows incrementing and not allow any other operations on it.

Reply:
> can you point me to "the sort and double collect for spans"? (Also, this comment was resolved, what was the outcome?)

This got taken out, since as long as we can get the guarantee that Id sequence corresponds to the sequence spans are entered then we don't need to include `time_entered`. Originally I was sorting on `time_entered`, which is not in the version of the span that gets sent over the wire. So I had to collect the spans from the subscriber, sort them and then map over them again to convert them to the type we send over the wire. However, no I just sort the spans by Id with api-sidecar.

Reply:
Let me know what to do with the ordering